Mathlib/AlgebraicGeometry/EllipticCurve/Weierstrass.lean

```diff
@@ -0,0 +1,995 @@
+
+ * `WeierstrassCurve.twoTorsionPolynomial_disc`: the discriminant of a Weierstrass curve is a
+    constant factor of the cubic discriminant of its 2-torsion polynomial.
+ * `WeierstrassCurve.nonsingular_of_Δ_ne_zero`: a Weierstrass curve is nonsingular at every point
+    if its discriminant is non-zero.
+ * `WeierstrassCurve.CoordinateRing.instIsDomainCoordinateRing`: the coordinate ring of a
+    Weierstrass curve is an integral domain.
+ * `WeierstrassCurve.CoordinateRing.degree_norm_smul_basis`: the degree of the norm of an element
+    in the coordinate ring in terms of the power basis.
+ * `EllipticCurve.nonsingular`: an elliptic curve is nonsingular at every point.
+ * `EllipticCurve.variableChange_j`: the j-invariant of an elliptic curve is invariant under an
+    admissible linear change of variables.
+
+## Implementation notes
+
+The definition of elliptic curves in this file makes sense for all commutative rings `R`, but it
+only gives a type which can be beefed up to a category which is equivalent to the category of
+elliptic curves over the spectrum $\mathrm{Spec}(R)$ of `R` in the case that `R` has trivial Picard
+group $\mathrm{Pic}(R)$ or, slightly more generally, when its 12-torsion is trivial. The issue is
+that for a general ring `R`, there might be elliptic curves over $\mathrm{Spec}(R)$ in the sense of
+algebraic geometry which are not globally defined by a cubic equation valid over the entire base.
+
+## References
+
+ * [N Katz and B Mazur, *Arithmetic Moduli of Elliptic Curves*][katz_mazur]
+ * [P Deligne, *Courbes Elliptiques: Formulaire (d'après J. Tate)*][deligne_formulaire]
+ * [J Silverman, *The Arithmetic of Elliptic Curves*][silverman2009]
+
+## Tags
+
+elliptic curve, weierstrass equation, j invariant
+-/
+
+-- porting note: replaced `map_one`, `map_bit0`, and `map_bit1` with `map_ofNat`
+local macro "map_simp" : tactic =>
+  `(tactic| simp only [map_ofNat, map_neg, map_add, map_sub, map_mul, map_pow])
+
+universe u v w
+
+variable {R : Type u}
+
+/-! ## Weierstrass curves -/
+
+/-- A Weierstrass curve $Y^2 + a_1XY + a_3Y = X^3 + a_2X^2 + a_4X + a_6$ with parameters $a_i$. -/
+@[ext]
+structure WeierstrassCurve (R : Type u) where
+  (a₁ a₂ a₃ a₄ a₆ : R)
+#align weierstrass_curve WeierstrassCurve
+
+namespace WeierstrassCurve
+
+/-- The `a₁` coefficient of a Weierstrass curve. -/
+add_decl_doc a₁
+
+/-- The `a₂` coefficient of a Weierstrass curve. -/
+add_decl_doc a₂
+
+/-- The `a₃` coefficient of a Weierstrass curve. -/
+add_decl_doc a₃
+
+/-- The `a₄` coefficient of a Weierstrass curve. -/
+add_decl_doc a₄
+
+/-- The `a₆` coefficient of a Weierstrass curve. -/
+add_decl_doc a₆
+
+instance [Inhabited R] : Inhabited <| WeierstrassCurve R :=
+  ⟨⟨default, default, default, default, default⟩⟩
+#align weierstrass_curve.inhabited WeierstrassCurve.instInhabitedWeierstrassCurve
+
+variable [CommRing R] (W : WeierstrassCurve R)
+
+section Quantity
+
+/-! ### Standard quantities -/
+
+-- porting note: removed `@[simp]`
```

Comment:
We need to say why are each note (does simp already do this, are you avoiding a simpNF linter error, etc?)